zerver/tests/test_message_edit.py

```diff
@@ -1654,3 +1654,168 @@ def test_user_group_mention_restrictions_while_editing(self) -> None:
+        self.assertEqual(result_content["result"], "success")
+        self.assert_length(result_content["detached_files"], 0)
+
+        # Try editing second message, the only reference to the attachment now
+        result = self.client_patch(
+            f"/json/messages/{msg_id}",
+            {
+                "content": "Try editing a message with an attachment",
+            },
+        )
+        result_content = orjson.loads(result.content)
+        self.assertEqual(result_content["result"], "success")
+        self.assert_length(result_content["detached_files"], 1)
+        actual_path_id_set = (
+            CONST_UPLOAD_PATH_PREFIX + result_content["detached_files"][0]["path_id"]
+        )
+
+        self.assertEqual(actual_path_id_set, file1)
+
+        result = self.client_patch(
+            f"/json/messages/{msg_id}",
+            {
+                "content": "Try editing a message with no attachments",
+            },
+        )
+        result_content = orjson.loads(result.content)
+        self.assertEqual(result_content["result"], "success")
+        self.assert_length(result_content["detached_files"], 0)
+
+    def test_remove_another_user_attachment_while_editing(self) -> None:
+        # Try editing a message and removing an linked attachment that's
+        # uploaded by another user. Users should not be able to detach another
+        # user's attachments.
+
+        user_profile = self.example_user("hamlet")
+        file1 = self.create_attachment_helper(user_profile)
+
+        content = f"Init message [attachment1.txt]({file1})"
+
+        # Send a message with attachment using another user profile.
+        msg_id = self.send_stream_message(
+            user_profile,
+            "Denmark",
+            topic_name="editing",
+            content=content,
+        )
+        self.check_message(msg_id, topic_name="editing", content=content)
+        attachments = Attachment.objects.filter(messages__in=[msg_id])
+        self.assert_length(attachments, 1)
+
+        # Send a message referencing to the attachment uploaded by another user.
+        self.login("iago")
+        msg_id = self.send_stream_message(
+            self.example_user("iago"),
+            "Denmark",
+            topic_name="editing",
+            content=content,
+        )
+        self.check_message(msg_id, topic_name="editing", content=content)
+        attachments = Attachment.objects.filter(messages__in=[msg_id])
+        self.assert_length(attachments, 1)
+
+        # Try editing the message and removing the reference to the attachment.
+        result = self.client_patch(
+            f"/json/messages/{msg_id}",
+            {
+                "content": "Try editing a message with an attachment uploaded by another user",
+            },
+        )
+        result_content = orjson.loads(result.content)
+        self.assertEqual(result_content["result"], "success")
+        self.assert_length(result_content["detached_files"], 0)
```

Comment:
I'm not sure this test is verifying anything useful. I think one would need that very first message to have been deleted in order for the `detached_files` list to mean anything; the file is not detached because the first message still exists.

Reply:
Ahh, it looks like the next test is that ... so I'm not sure what this test is here for.

Reply:
This test is to ensure that one doesn't get another person's attachments as detached_files.